### DISCUSSION TITLE
Use form state to dynamically render elements

### DISCUSSION THREAD

**Discussion by dMaisnam** [react-hook-form/react-hook-form#8283] [Q&A] [answered]
Hello. I am working on a CMS web app. I have a default data provided in the schema below
```js
const data = {
  title: "",
  ....., // remaining input goes here
  category: [
    { id: 1, name: "News", subcategories: null },
    { id: 2, name: "Tech", subcategories: [
      { id: 1, name: "Blockchain" },
      // and so on
    ]}
  ]
}
```
I have an example [codesandbox](https://codesandbox.io/s/react-form-2mup06?file=/src/App.tsx) implementation of this. I want to use the values `category` and `subcategory` from the form select inputs so that I can loop over to display them as the options. 
```js
<div className="form-group">
  <label htmlFor="category">Category</label>
  <select id="category" {...register("category")}>
    {/* Map over `category` array here to list option of categories */}
  </select>
</div>
```
I would also like to dynamically render the subcategories based on whether the category has subcategories or not. For this I would like to achieve this without using state (`useState`, `redux`, etc.) if possible. 
```js
const { register, handleSubmit } = useForm({
  defaultValues: {
    title: data.title,
    excerpt: data.excerpt,
    content: data.content,
    category: data.category.map(cat => cat.name),
    // subcategory: data.category.
  }
})
```
How do I go about to achieve this? Is my default schema defined above not a good design ? I am doing in such a way so that I can also use the same logic in the form update process. Thank you for reading.

**Comment by Moshyfawn** [comment, marked as answer]
First off, you can't watch a value before initializing the form. There's no category form value until you at least define your form defaults and `useForm`.
```js
const category = watch(["category"])
useForm({
  defaultValues: {
    subcategory: data.category
      .filter((cat) => cat.id === **category.id**)
      .subcategories.map((subcat) => subcat.name)
})
```

Also, it looks like your entities types _(`Category`, `Article`)_ have misaligned types with your desired form values shape is. Like 
```
type ArticleDataProps = {
  category: CategoryProps
}
```
But your initial `data` and the form's default `category` values are an `array`
```js
const data = {
  category: [
    { id: 1, name: "News", subcategories: null },
  ]
}

const defaultValues = {
  category: data.category.map((cat) => cat.name)
}
```

**Reply by Moshyfawn** [reply]
I see that you've updated your codesanbox now :P

**Reply by Moshyfawn** [reply]
If you pass your ArticleDataProps into the `useForm`'s generic, you'll see that your data shapes don't match
```ts
  const { register, handleSubmit } = useForm<ArticleDataProps>({
    defaultValues: {
      title: data.title,
      excerpt: data.excerpt,
      content: data.content,
      // INFO: TS issue here
      category: data.category.map((cat) => cat.name),
      // INFO: TS issue here
      subcategory: null
    }
  })
```

**Reply by Moshyfawn** [reply]
Seems like this is the type of form's values you're going for
```ts
interface FormValues {
  title: ArticleDataProps["title"];
  excerpt: ArticleDataProps["excerpt"];
  content: ArticleDataProps["content"];
  category: CategoryProps["name"][];
  subcategory: SubcategoryProps["name"][] | null;
}
```

**Reply by dMaisnam** [reply]
Thanks for the response. I have an idea what you are saying. I will further look into this

**Reply by Moshyfawn** [reply]
Or not, 'cuz you want to have only one category selected at a time :P

Then your `defaultValues` should reflect that.

Here you go: [something along the lines of this example](https://codesandbox.io/s/react-form-forked-1gw83c?file=%2Fsrc%2Fcomponents%2FCreateArticle.tsx) would suffices. Obvsly, needs so clean up and logic handling

**Reply by Moshyfawn** [reply]
Notice, that I've opt-in to use an empty string as a `defaultValue` for both select inputs _(category, subcategory)_ to have a default "empty state" value selected initially

**Reply by dMaisnam** [reply]
Yes I think I am getting what you mean. Thanks for the help. Much appreciated.
